Start Live Text image analysis when Find-in-Page is performed on MacOS Safari

danae404 opened the pull request:
#### 86cccca6075ebd2e357c6ab410881af35727d970
<pre>
Start Live Text image analysis when Find-in-Page is performed on MacOS Safari
<a href="https://bugs.webkit.org/show_bug.cgi?id=241877">https://bugs.webkit.org/show_bug.cgi?id=241877</a>
rdar://95726585

Reviewed by Wenson Hsieh.

Test:
FindInPAgeAPI.FindAPITextInImage

This patch makes it so that when Find-in-Page is performed, text within images
On the page is included in the results. This feature is placed behind a WebKit
internal feature flag.

Added a test that checks that an image analysis request is made after doing
Find-in-page.

* Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml:

Added internal feature flag &quot;ImageAnalysisDuringFindInPageEnabled&quot;.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::analyzeImagesForFindInPage):

Helper function that is called from FindController::FindString. If flag is
turned on, images are analyzed.

* Source/WebCore/page/Page.h:
* Source/WebKit/WebProcess/WebPage/FindController.cpp:
(WebKit::FindController::findString):

Now starts image analysis before beginning find-in-page.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/FindInPageAPI.mm:
(FindInPageAPI::createWebViewWithImageAnalysisDuringFindInPageEnabled):
(FindInPageAPI::makeImageAnalysisRequestSwizzler):
(FindInPageAPI:makeFakeRequest):
(FindInPageAPI:processRequestWithResults):

Canonical link: <a href="https://commits.webkit.org/252515@main">https://commits.webkit.org/252515@main</a>
</pre>

pxlcoder commented on Source/WebKit/WebProcess/WebPage/FindController.cpp:
```diff
@@ -237,6 +238,13 @@ void FindController::findString(const String& string, OptionSet<FindOptions> opt
+        Document* document = m_webPage->mainWebFrame().coreFrame()->document();
+        if (document)
+            corePage->imageAnalysisQueue().enqueueAllImages(*document, { }, { });
+    }
```
Can we move this logic into `Page::findTextMatches` or another helper method to be called from `Page::findTextMatches`?

As it stands, we have a couple of places where `Page::findTextMatches` is used to retrieve results. One for WebKit-specific API/SPI, one for AppKit (`NSTextFinder`), and one for UIKit (`UIFindInteraction`). This would only cover the first case.

whsieh replied:
Once you move the code into `findTextMatches`, I would also write this as something like:

```
if (settings().liveTextForFindInPageEnabled()) {
    if (RefPtr mainDocument = m_mainFrame->document())
        imageAnalysisQueue().enqueueAllImages(*mainDocument, { }, { });
}
```

On a related note, is this only invoked once at the beginning of the Find in Page invocation? (We should either ensure that this new logic is inside an entry point that's only called once at the beginning in all three scenarios, or we should adjust `enqueueAllImages` to be idempotent).

(Edit: perhaps something like this?)

```
if (settings().liveTextForFindInPageEnabled()) {
    if (RefPtr mainDocument = m_mainFrame->document(); mainDocument && !m_imageAnalysisQueue)
        imageAnalysisQueue().enqueueAllImages(*mainDocument, { }, { });
}
```

danae404 replied:
> Can we move this logic into `Page::findTextMatches` or another helper method to be called from `Page::findTextMatches`?

Unless I'm mistaken, the case that my original patch covers is no longer covered if I move the code to `Page::findTextMatches`. From `FindController::findString`, `findTextMatches` is only called `if (selectedFrame->selection().selectionBounds().isEmpty())`

pxlcoder replied:
We talked about this offline, and have decided to try "another helper method to be called from `Page::findTextMatches`", and additionally call the helper from `Page::findString`.

The difference between the WebKit-specific case and the AppKit/UIKit cases, is that WebKit only searches for the next match following the current selection, whereas AppKit/UIKit expect all results up front.

`findPlainText` in `TextIterator` is the common point between the codepaths, but not sure that the logic should live in `TextIterator`, hence the helper method approach.

pxlcoder commented on Source/WebCore/page/Page.cpp:
```diff
@@ -809,8 +809,22 @@ bool Page::findString(const String& target, FindOptions options, DidWrap* didWra
 }
 
+#if ENABLE(IMAGE_ANALYSIS)
+void Page::analyzeImagesForFindInPage()
```
Does calling this method multiple times re-trigger image analysis, or will it only run the logic once?

whsieh replied:
The added `!m_imageAnalysisQueue` check should make this idempotent. (It'll also do nothing if the image analysis queue has already been created to service visual translation when translating webpages, which is expected).

whsieh commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/FindInPageAPI.mm:
```diff
@@ -33,6 +33,13 @@
+#import "ImageAnalysisTestingUtilities.h"
+#import "InstanceMethodSwizzler.h"
+#import <WebKit/_WKInternalDebugFeature.h>
+#import <pal/spi/cocoa/VisionKitCoreSPI.h>
```
Nit - I think you can leave out this import, if you're already importing `VisionKitCoreSoftLink.h`

danae404 replied:
Fixed!

whsieh commented on Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml:
```diff
@@ -572,6 +572,18 @@ LiveRangeSelectionEnabled:
     WebCore:
       default: false
 
+LiveTextForFindInPageEnabled:
```
Nit - w.r.t. setting names, I've been largely avoiding names like "Live Text" that are directly tied to features on Apple platforms (though currently, it's true that it's only relevant on Apple platforms). What do you think about something slightly more general, like `TextRecognitionDuringFindInPageEnabled` or `ImageAnalysisDuringFindInPageEnabled`?

danae404 replied:
Fixed!

whsieh reviewed (approved):
Seems like a good start to me (just a couple of minor comments!)

danae404 committed:
Start Live Text image analysis when Find-in-Page is performed on MacOS Safari
https://bugs.webkit.org/show_bug.cgi?id=241877
rdar://95726585

Reviewed by Wenson Hsieh.

Test:
FindInPAgeAPI.FindAPITextInImage

This patch makes it so that when Find-in-Page is performed, text within images
On the page is included in the results. This feature is placed behind a WebKit
internal feature flag.

Added a test that checks that an image analysis request is made after doing
Find-in-page.

* Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml:

Added internal feature flag "ImageAnalysisDuringFindInPageEnabled".

* Source/WebCore/page/Page.cpp:
(WebCore::Page::analyzeImagesForFindInPage):

Helper function that is called from FindController::FindString. If flag is
turned on, images are analyzed.

* Source/WebCore/page/Page.h:
* Source/WebKit/WebProcess/WebPage/FindController.cpp:
(WebKit::FindController::findString):

Now starts image analysis before beginning find-in-page.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/FindInPageAPI.mm:
(FindInPageAPI::createWebViewWithImageAnalysisDuringFindInPageEnabled):
(FindInPageAPI::makeImageAnalysisRequestSwizzler):
(FindInPageAPI:makeFakeRequest):
(FindInPageAPI:processRequestWithResults):

Canonical link: https://commits.webkit.org/252515@main
